### PR TITLE
Add additional services specific controls. Closes #53

### LINKED ISSUE #53: Add additional services specific controls
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Priyanka-Chatterjee-2000**
Controls and queries added:

1. [x] ec2_ami_copy_encryption_enabled 
2. [x] ec2_ami_copy_encrypted_with_kms_cmk 
3. [x] ec2_ami_encryption_enabled 
4. [x] apigateway_rest_api_create_before_destroy_enabled
5. [x] apigateway_deployment_create_before_destroy_enabled
6. [x] apigateway_method_settings_cache_enabled
7. [x] apigateway_method_settings_cache_encryption_enabled 
8. [x] apigateway_method_settings_data_trace_enabled
9. [x] autoscaling_group_uses_launch_template
10. [x] autoscaling_group_tagging_enabled
11. [x] backup_vault_encryption_at_rest_enabled
12. [x] ebs_snapshot_copy_encrypted_with_kms_cmk 
13. [x] ec2_instance_user_data_no_secrets
14. [x] ec2_ami_imagebuilder_component_encrypted_with_kms_cmk 
15. [x] ec2_ami_imagebuilder_distribution_configuration_encrypted_with_kms_cmk
16. [x] ec2_ami_imagebuilder_image_recipe_encrypted_with_kms_cmk 
17. [x] ec2_launch_configuration_ebs_encryption_check 
18. [x] ec2_instance_ebs_encryption_check
19. [x] ec2_launch_template_metadata_hop_limit_check
20. [x] ec2_launch_configuration_metadata_hop_limit_check
21. [x] rds_db_cluster_encrypted_with_kms_cmk
22. [x] s3_bucket_abort_incomplete_multipart_upload_enabled
23. [x] s3_bucket_block_public_policy_enabled
24. [x] s3_bucket_object_encrypted_with_kms_cmk
25. [x] s3_bucket_ignore_public_acls_enabled
26. [x] s3_bucket_object_copy_encrypted_with_kms_cmk
27. [x] sqs_queue_policy_no_action_star
28. [x] sqs_queue_policy_no_principal_star
29. [x] ssm_document_prohibit_public_access
30. [x] ssm_parameter_encrypted_with_kms_cmk
31. [x] sagemaker_domain_encrypted_with_kms_cmk
32. [x] sagemaker_notebook_instance_in_vpc
33. [x] sagemaker_notebook_instance_root_access_disabled
34. [x] sfn_state_machine_execution_history_logging_enabled
35. [x] sfn_state_machine_xray_tracing_enabled
36. [x] timestream_database_encrypted_with_kms_cmk
37. [x] waf_web_acl_rule_attached
38. [x] waf_regional_web_acl_rule_attached
39. [x] wafv2_web_acl_rule_attached
40. [x] waf_web_acl_logging_enabled
41. [x] wafregional_web_acl_logging_enabled
42. [x] waf_web_acl_rule_with_action
43. [x] wafregional_web_acl_rule_with_action

Controls and queries removed:

1. [x] ebs_attached_volume_encryption_enabled

### CODE REVIEW

**Comment by rajlearner17** [comment]
@khushboo9024 @Priyanka-Chatterjee-2000 
Better we remove this control from EBS, i.e. `Attached EBS volumes should have encryption enabled [`ebs_attached_volume_encryption_enabled`], which uses the same query, i.e. [ebs_volume_encryption_at_rest_enabled](https://hub-steampipe-io-git-staging-turbot.vercel.app/mods/turbot/terraform_aws_compliance/queries/ebs_volume_encryption_at_rest_enabled) as it in [EBS volumes should have encryption enabled](https://hub-steampipe-io-git-staging-turbot.vercel.app/mods/turbot/terraform_aws_compliance/controls/control.ebs_volume_encryption_at_rest_enabled?context=benchmark.ebs)

Over to that, we have no direct provision to enable EBS attachment in TF; hence this control `ebs_attached_volume_encryption_enabled` can be removed to avoid confusion.

**Review by rajlearner17 — APPROVED** [review]
LGTM

File: controls/apigateway.sp [hunk 1]
```diff
@@ -67,3 +73,51 @@ control "apigateway_stage_logging_enabled" {
...
+  title       = "API Gateway Method Settings should have cache encrypted"
+  description = "This control checks whether AWS API Gateway Method Settings has cache encrypted. It is recommended to enable cache encryption for all methods in API Gateway."
+  query       = query.apigateway_method_settings_cache_encryption_enabled
+
+  tags = local.apigateway_compliance_common_tags
+}
+
+control "apigateway_method_settings_data_trace_enabled" {
+  title       = "API Gateway Method Settings should have data trace enabled"
+  description = "This control checks whether AWS API Gateway Method Settings has data trace enabled. It is recommended to disable data trace for all methods in API Gateway."
```
Comment (madhushreeray30) [inline controls/apigateway.sp:111]: What is the query logic here? Should it be enabled or disabled? The query prompts it to be disabled while the title and description are contradicting.
I believe it should be `enabled` @Priyanka-Chatterjee-2000 please update it accordingly.

File: query/backup.sp [hunk 2]
```diff
@@ -1,3 +1,24 @@
+query "backup_vault_encryption_at_rest_enabled" {
+  sql = <<-EOQ
+    select
+      type || ' ' || name as resource,
+      case
+        when (arguments -> 'kms_key_arn') is null then 'alarm'
+        else 'ok'
+      end status,
+      name || case
+        when (arguments -> 'kms_key_arn') is null then ' encryption at rest not enabled'
```
Comment (madhushreeray30) [inline query/backup.sp:10]:
```suggestion
        when (arguments -> 'kms_key_arn') is null then ' encryption at rest disabled'
```

File: query/ec2.sp [hunk 3]
```diff
@@ -195,3 +195,238 @@ query "ec2_instance_uses_imdsv2" {
...
+query "ec2_ami_imagebuilder_component_encrypted_with_kms_cmk" {
+  sql = <<-EOQ
+    select
+      type || ' ' || name as resource,
+      case
+        when (arguments ->> 'kms_key_id') is null then 'alarm'
+        else 'ok'
+      end as status,
+      name || case
+        when (arguments ->> 'kms_key_id') is null then ' is not encrypted with customer-managed CMK.'
```
Comment (madhushreeray30) [inline query/ec2.sp:233]:
```suggestion
        when (arguments ->> 'kms_key_id') is null then ' is not encrypted with CMK'
```

File: query/ec2.sp [hunk 4]
```diff
@@ -195,3 +195,238 @@ query "ec2_instance_uses_imdsv2" {
...
+        else ' is encrypted with customer-managed CMK.'
```
Comment (madhushreeray30) [inline query/ec2.sp:234]:
```suggestion
        else ' is encrypted with CMK'
```

File: query/ec2.sp [hunk 5]
```diff
@@ -195,3 +195,238 @@ query "ec2_instance_uses_imdsv2" {
...
+query "ec2_ami_imagebuilder_distribution_configuration_encrypted_with_kms_cmk" {
+  sql = <<-EOQ
+    select
+      type || ' ' || name as resource,
+      case
+        when (arguments -> 'distribution' -> 'ami_distribution_configuration' ->> 'kms_key_id') is null then 'alarm'
+        else 'ok'
+      end as status,
+      name || case
+        when (arguments -> 'distribution' -> 'ami_distribution_configuration' ->> 'kms_key_id') is null then ' is not encrypted with customer-managed CMK'
```
Comment (madhushreeray30) [inline query/ec2.sp:254]:
```suggestion
        when (arguments -> 'distribution' -> 'ami_distribution_configuration' ->> 'kms_key_id') is null then ' is not encrypted with CMK'
```

File: query/ec2.sp [hunk 6]
```diff
@@ -195,3 +195,238 @@ query "ec2_instance_uses_imdsv2" {
...
+        else ' is encrypted with customer-managed CMK'
```
Comment (madhushreeray30) [inline query/ec2.sp:255]:
```suggestion
        else ' is encrypted with CMK'
```

File: query/ec2.sp [hunk 7]
```diff
@@ -195,3 +195,238 @@ query "ec2_instance_uses_imdsv2" {
...
+query "ec2_ami_imagebuilder_image_recipe_encrypted_with_kms_cmk" {
+  sql = <<-EOQ
+    select
+      type || ' ' || name as resource,
+      case
+        when (arguments -> 'block_device_mapping' -> 'ebs' ->> 'kms_key_id') is null or (arguments -> 'block_device_mapping' -> 'ebs' ->> 'encrypted') <> 'true' or (arguments -> 'block_device_mapping' -> 'ebs' ->> 'encrypted') is null then 'alarm'
+        else 'ok'
+      end as status,
+      name || case
+        when (arguments -> 'block_device_mapping' -> 'ebs' ->> 'kms_key_id') is null or (arguments -> 'block_device_mapping' -> 'ebs' ->> 'encrypted') <> 'true' or (arguments -> 'block_device_mapping' -> 'ebs' ->> 'encrypted') is null then ' is not encrypted with customer-managed CMK'
```
Comment (madhushreeray30) [inline query/ec2.sp:275]:
```suggestion
        when (arguments -> 'block_device_mapping' -> 'ebs' ->> 'kms_key_id') is null or (arguments -> 'block_device_mapping' -> 'ebs' ->> 'encrypted') <> 'true' or (arguments -> 'block_device_mapping' -> 'ebs' ->> 'encrypted') is null then ' is not encrypted with CMK'
```

File: query/ec2.sp [hunk 8]
```diff
@@ -195,3 +195,238 @@ query "ec2_instance_uses_imdsv2" {
...
+        else ' is encrypted with customer-managed CMK'
```
Comment (madhushreeray30) [inline query/ec2.sp:276]:
```suggestion
        else ' is encrypted with CMK'
```

File: query/ec2.sp [hunk 9]
```diff
@@ -195,3 +195,238 @@ query "ec2_instance_uses_imdsv2" {
...
+  sql = <<-EOQ
+    select
+      type || ' ' || name as resource,
+      case
+        when (arguments -> 'metadata_options' ->> 'http_put_response_hop_limit')::int > 1 then 'alarm'
+        else 'ok'
+      end as status,
+      name || case
+        when (arguments -> 'metadata_options' ->> 'http_put_response_hop_limit')::int > 1 then ' metadata response hop limit value is greater than 1'
+        else ' metadata response hop limit value is not greater than 1'
```
Comment (madhushreeray30) [inline query/ec2.sp:297]:
```suggestion
        else ' metadata response hop limit value is not less than 1'
```

File: query/ec2.sp [hunk 10]
```diff
@@ -195,3 +195,238 @@ query "ec2_instance_uses_imdsv2" {
...
```
Comment (madhushreeray30) [inline query/ec2.sp:318]:
```suggestion
        else ' metadata response hop limit value is not less than 1'
```

File: query/rds.sp [hunk 11]
```diff
@@ -234,6 +234,27 @@ query "rds_db_instance_and_cluster_no_default_port" {
...
+query "rds_db_cluster_encrypted_with_kms_cmk" {
+  sql = <<-EOQ
+    select
+      type || ' ' || name as resource,
+      case
+        when (arguments -> 'kms_key_id') is null then 'alarm'
+        else 'ok'
+      end status,
+      name || case
+        when (arguments -> 'kms_key_id') is null then ' does not use KMS encryption'
```
Comment (madhushreeray30) [inline query/rds.sp:246]:
```suggestion
        when (arguments -> 'kms_key_id') is null then ' is encrypted with KMS CMK'
```

File: query/rds.sp [hunk 12]
```diff
@@ -234,6 +234,27 @@ query "rds_db_instance_and_cluster_no_default_port" {
...
+        else ' uses KMS encryption'
```
Comment (madhushreeray30) [inline query/rds.sp:247]:
```suggestion
        else ' is not encrypted with KMS CMK'
```

File: query/redshift.sp [hunk 13]
```diff
@@ -203,4 +203,25 @@ query "redshift_cluster_prohibit_public_access" {
...
+query "redshift_cluster_no_default_database_name" {
+  sql = <<-EOQ
+    select
+      type || ' ' || name as resource,
+      case
+        when (arguments ->> 'database_name') is not null then 'ok'
+        else 'alarm'
+      end status,
+      name || case
+        when (arguments ->> 'database_name') is not null then ' has database name defined'
```
Comment (madhushreeray30) [inline query/redshift.sp:217]:
```suggestion
        when (arguments ->> 'database_name') is not null then ' database name defined'
```

File: query/s3.sp [hunk 14]
```diff
@@ -247,4 +247,118 @@ query "s3_public_access_block_account" {
...
+        when (arguments -> 'block_public_policy') is null then ' block_public_acls not defined'
+        when (arguments ->> 'block_public_policy')::boolean then ' block_public_policy enabled'
+        else ' block_public_policy disabled'
+      end || '.' as reason
+      ${local.common_dimensions_sql}
+    from
+      terraform_resource
+    where
+      type = 'aws_s3_bucket_public_access_block';
+  EOQ
```
Comment (madhushreeray30) [inline query/s3.sp:270]: I believe we should have a skip condition here.

File: query/s3.sp [hunk 15]
```diff
@@ -247,4 +247,118 @@ query "s3_public_access_block_account" {
...
+        when (arguments -> 'ignore_public_acls') is null then ' ignore_public_acls not defined'
+        when (arguments ->> 'ignore_public_acls')::boolean then ' ignore_public_acls enabled'
+        else ' ignore_public_acls disabled'
+      end || '.' as reason
+      ${local.common_dimensions_sql}
+    from
+      terraform_resource
+    where
+      type = 'aws_s3_bucket_public_access_block';
+  EOQ
```
Comment (madhushreeray30) [inline query/s3.sp:311]: I believe we should have a skip condition here.

File: query/waf.sp [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+      name || case
+        when (arguments -> 'rules') is not null then ' has rule(s) attached'
+        else ' has no attached rules'
+      end || '.' reason
+      ${local.tag_dimensions_sql}
+      ${local.common_dimensions_sql}
+    from
+      terraform_resource
+    where
+      type = 'aws_waf_web_acl'
```
Comment (madhushreeray30) [inline query/waf.sp:18]:
```suggestion
      type = 'aws_waf_web_acl';
```

File: query/waf.sp [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+      name || case
+        when (arguments -> 'logging_configuration' ->> 'log_destination') is not null then ' logging enabled'
+        else ' logging disabled'
+      end || '.' reason
+      ${local.tag_dimensions_sql}
+      ${local.common_dimensions_sql}
+    from
+      terraform_resource
+    where
+      type = 'aws_waf_web_acl'
```
Comment (madhushreeray30) [inline query/waf.sp:60]:
```suggestion
      type = 'aws_waf_web_acl';
```

File: query/waf.sp [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+      name || case
+        when (arguments -> 'logging_configuration' ->> 'log_destination' ) is not null then ' logging enabled'
+        else ' logging disabled'
+      end || '.' reason
+      ${local.tag_dimensions_sql}
+      ${local.common_dimensions_sql}
+    from
+      terraform_resource
+    where
+      type = 'aws_wafregional_web_acl'
```
Comment (madhushreeray30) [inline query/waf.sp:81]:
```suggestion
      type = 'aws_wafregional_web_acl';
```

**Review by madhushreeray30 — COMMENTED** [review]
@Priyanka-Chatterjee-2000 can you please have a look at the review comments, thanks!

**Review by madhushreeray30 — COMMENTED** [review]
@Priyanka-Chatterjee-2000 can you please have a look at the review comments, thanks!

File: controls/apigateway.sp (hunk 1 above)
Comment (Priyanka-Chatterjee-2000) [reply controls/apigateway.sp:111]: No it should be disabled only.

File: query/s3.sp (hunk 14 above)
Comment (khushboo9024) [reply query/s3.sp:270]: @madhushreeray30 this can't be skipped , we are checking `aws_s3_bucket_public_access_block` terraform resource.

**Review by madhushreeray30 — APPROVED** [review]

**Pull request merged**